docs/proposals/templates.md

```diff
@@ -0,0 +1,350 @@
+expanded.
+
+Additional example ("openshift" syntax):
+```
+kind: "TemplateDeclaration"
+apiVersion: "v1"
+metadata: 
+  name: "mongodb-ephemeral"
+  annotations: 
+    description: "Provides a MongoDB database service"
+templateKind: openshift
+parameters: 
+  - 
+    name: "DATABASE_SERVICE_NAME"
+    description: "Database service name"
+    value: "mongodb"
+    required: true
+    type: "string"
+  - 
+    name: "MONGODB_USER"
+    description: "Username for MongoDB user that will be used for accessing the database"
+    required: true
+    type: "string"
+  - 
+    name: "MONGODB_PASSWORD"
+    description: "Password for the MongoDB user"
+    required: true
+    type: "string"
+  - 
+    name: "MONGODB_DATABASE"
+    description: "Database name"
+    value: "sampledb"
+    required: true
+    type: "string"
+  - 
+    name: "REPLICA_COUNT"
+    description: "Number of mongo replicas to run"
+    value: "1"
+    required: true
+    type: "int"
+body: 
+  objects: 
+    - 
+      kind: "Service"
+      apiVersion: "v1"
+      metadata: 
+        name: "$(DATABASE_SERVICE_NAME)"
+      spec: 
+        ports: 
+          - 
+            name: "mongo"
+            protocol: "TCP"
+            targetPort: 27017
+        selector: 
+          name: "$(DATABASE_SERVICE_NAME)"
+    - 
+      kind: "ReplicationController"
+      apiVersion: "v1"
+      metadata: 
+        name: "$(DATABASE_SERVICE_NAME)"
+      spec: 
+        replicas: "$(REPLICA_COUNT)"
+        selector: 
+          name: "$(DATABASE_SERVICE_NAME)"
+        template: 
+          metadata: 
+            creationTimestamp: null
+            labels: 
+              name: "$(DATABASE_SERVICE_NAME)"
+          spec: 
+            containers: 
+              - 
+                name: "mongodb"
+                image: "docker.io/centos/mongodb-26-centos7"
+                ports: 
+                  - 
+                    containerPort: 27017
+                    protocol: "TCP"
+                env: 
+                  - 
+                    name: "MONGODB_USER"
+                    value: "${MONGODB_USER}"
+                  - 
+                    name: "MONGODB_PASSWORD"
+                    value: "${MONGODB_PASSWORD}"
+                  - 
+                    name: "MONGODB_DATABASE"
+                    value: "${MONGODB_DATABASE}"
+```
+
+## Template Expansion API
+
+A server side endpoint for expanding templates is proposed.  The endpoint will inspect the `templateKind` for the template object
+being posted to the endpoint, and then use that information to determine the appropriate template expansion plugin to invoke.
+
+A template expansion plugin is expected to return a `List` of valid k8s objects, which the template processing endpoint
+will then return to the original caller.  This represents the primary contract for templates:  A template api object 
+envelope containing arbitrary content is converted into a list of k8s objects.
```

Comment:
Why return the expansion results to the caller? If that's what the user wants, they don't need to use a k8s cluster to do it. Instead, they can just run tools on their clients.

The whole point of expanding templates server side, to my thinking, is to capture the templates and the expansion on the server, so that the cluster knows what was instantiated in it, and callers can discover that information.

Furthermore, if the plugin, which I've been called an object processor, returns the results to the api server, then the api server, can recursively process them again, thereby implementing hierarchy, which is a very powerful topology tool.